### PR TITLE
SONAR-6610 WS custom_measures/create create a custom measure

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+    action.createParam(PARAM_PROJECT_ID)
+      .setDescription("Project id")
+      .setExampleValue("ce4c03d6-430f-40a9-b777-ad877c00aa4d");
+
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("Project key")
+      .setExampleValue("org.apache.hbas:hbase");
+
+    action.createParam(PARAM_METRIC_ID)
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: Shouldn't we allow metric key also, for "public" WS use?

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      .setRequired(true)
+      .setDescription("Metric id")
+      .setExampleValue("16");
+
+    action.createParam(PARAM_VALUE)
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: This parameter could have a description of the supported formats. Especially for BOOL (since the WS convention allows ``true``, ``false``, ``yes``, ``no``) and WORK_DUR.

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+          boolean booleanValue = request.mandatoryParamAsBoolean(PARAM_VALUE);
+          measure.setValue(booleanValue ? 1.0d : 0.0d);
+          break;
+        case INT:
+        case MILLISEC:
+          int intValue = request.mandatoryParamAsInt(PARAM_VALUE);
+          measure.setValue(intValue);
+          break;
+        case WORK_DUR:
+          long longValue = request.mandatoryParamAsLong(PARAM_VALUE);
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: Maybe we could use the ``1d2h3m`` format here, WDYT?

File: server/sonar-server/src/test/java/org/sonar/server/measure/custom/ws/CreateActionTest.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+    ws.newGetRequest(CustomMeasuresWs.ENDPOINT, CreateAction.ACTION)
+      .setParam(CreateAction.PARAM_PROJECT_ID, DEFAULT_PROJECT_UUID)
+      .setParam(CreateAction.PARAM_METRIC_ID, "whatever-id")
+      .setParam(CreateAction.PARAM_VALUE, "custom-measure-free-text")
+      .execute();
+  }
+
+  @Test
+  public void fail_when_nor_project_id_nor_project_key_provided() throws Exception {
```
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/measure/custom/ws/CreateActionTest.java]: Maybe add ``fail_when_both_project_id_and_project_key_provided`` - just a corner case.

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java (hunk 1 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: yes, it's done

File: server/sonar-server/src/main/java/org/sonar/server/util/MetricLevelTypeValidation.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  @Override
+  public String key() {
+    return PropertyType.METRIC_LEVEL.name();
+  }
+
+  @Override
+  public void validate(String value, @Nullable List<String> options) {
+    try {
+      Metric.Level.valueOf(value);
+    } catch (IllegalArgumentException e) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/util/MetricLevelTypeValidation.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    json.prop(FIELD_METRIC_ID, String.valueOf(metric.getId()));
+    json.prop(FIELD_METRIC_KEY, metric.getKey());
+    json.prop(FIELD_PROJECT_ID, component.uuid());
+    json.prop(FIELD_PROJECT_KEY, component.key());
+    json.prop(FIELD_DESCRIPTION, measure.getDescription());
+    json.prop(FIELD_VALUE, measureWithoutInternalFormatting);
+    json.endObject();
+  }
+
+  private void setMeasureValue(CustomMeasureDto measure, Request request, MetricDto metric) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: MAJOR The Cyclomatic Complexity of this method "setMeasureValue" is 16 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+      description.append("<li>");
+      description.append(String.format("%s - %s", metricType.description(), metricTypeWsDescription(metricType)));
+      description.append("</li>");
+    }
+    description.append("</ul>");
+
+    return description.toString();
+  }
+
+  private String metricTypeWsDescription(Metric.ValueType metricType) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: MAJOR The Cyclomatic Complexity of this method "metricTypeWsDescription" is 12 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+      return;
+    }
+
+    userSession.checkLoggedIn().checkProjectUuidPermission(UserRole.ADMIN, component.projectUuid());
+  }
+
+  private void checkMeasureDoesNotExistAlready(DbSession dbSession, ComponentDto component, MetricDto metric) {
+    int nbMeasuresOnSameMetricAndMeasure = dbClient.customMeasureDao().countByComponentIdAndMetricId(dbSession, component.uuid(), metric.getId());
+    if (nbMeasuresOnSameMetricAndMeasure > 0) {
+      throw new ServerException(HttpURLConnection.HTTP_CONFLICT, String.format("A measure already exists for project id '%s' and metric id '%d'", component.uuid(), metric.getId()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: MINOR Split this 181 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    }
+  }
+
+  private void writeMeasure(JsonWriter json, CustomMeasureDto measure, ComponentDto component, MetricDto metric, String measureWithoutInternalFormatting) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: MAJOR Make "writeMeasure" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+          typeValidations.validate(valueAsString, PropertyType.METRIC_WORK_DURATION.name(), null);
+          measure.setValue(durations.decode(valueAsString).toMinutes());
+          break;
+        case FLOAT:
+        case PERCENT:
+        case RATING:
+          typeValidations.validate(valueAsString, PropertyType.FLOAT.name(), null);
+          measure.setValue(Double.parseDouble(valueAsString));
+          break;
+        case LEVEL:
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CreateAction.java]: CRITICAL End this switch case with an unconditional break, continue, return or throw statement. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS128)

File: server/sonar-server/src/main/java/org/sonar/server/util/MetricWorkDurationTypeValidation.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  @Override
+  public String key() {
+    return PropertyType.METRIC_WORK_DURATION.name();
+  }
+
+  @Override
+  public void validate(String value, @Nullable List<String> options) {
+    try {
+      durations.decode(value);
+    } catch (IllegalArgumentException e) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/util/MetricWorkDurationTypeValidation.java]: CRITICAL Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* CRITICAL 3 critical
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/util/MetricLevelTypeValidation.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/util/MetricLevelTypeValidation.java]: won't fix, that's what I want

File: server/sonar-server/src/main/java/org/sonar/server/util/MetricWorkDurationTypeValidation.java (hunk 11 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/util/MetricWorkDurationTypeValidation.java]: won't fix, that's what I want

**Comment by jblievremont** [comment]
LGTM
